synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,253 @@
+        txn.execute(sql, [expiration_time])
+        session_ids = [r[0] for r in txn.fetchall()]
+
+        # Delete the corresponding completed credentials.
+        self.db.simple_delete_many_txn(
+            txn,
+            table="ui_auth_sessions_credentials",
+            column="session_id",
+            iterable=session_ids,
+            keyvalues={},
+        )
+
+        # Finally, delete the sessions.
+        self.db.simple_delete_many_txn(
+            txn,
+            table="ui_auth_sessions",
+            column="session_id",
+            iterable=session_ids,
+            keyvalues={},
+        )
+
+    async def mark_stage_complete(
+        self,
+        session_id: str,
+        stage_type: str,
+        identity: Union[str, bool, Dict[str, any]],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            identity: The identity authenticated by the stage.
+        """
+        await self.db.runInteraction(
+            "mark_stage_complete",
+            self._mark_stage_completed,
+            session_id,
+            stage_type,
+            identity,
+        )
+
+    def _mark_stage_completed(
+        self,
+        txn,
+        session_id: str,
+        stage_type: str,
+        identity: Union[str, bool, Dict[str, any]],
+    ):
+        # Add (or update) the results of the current stage to the database.
+        self.db.simple_upsert_txn(
+            txn,
+            table="ui_auth_sessions_credentials",
+            keyvalues={"session_id": session_id, "stage_type": stage_type},
+            values={"identity": json.dumps(identity)},
+        )
+        # Mark the session as still in use.
+        self.db.simple_update_one_txn(
+            txn,
+            table="ui_auth_sessions",
+            keyvalues={"session_id": session_id},
+            updatevalues={"last_used": self.hs.get_clock().time_msec()},
+        )
+
+    async def get_completed_stages(
+        self, session_id: str
+    ) -> Dict[str, Union[str, bool, Dict[str, Any]]]:
+        """
+        Retrieve the completed stages of a UI authentication session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            The completed stages mapped to the user which completed that stage.
+        Raises:
+            StoreError: if the session is not found.
+        """
+        results = {}
+        for row in await self.db.simple_select_list(
+            table="ui_auth_sessions_credentials",
+            keyvalues={"session_id": session_id},
+            retcols=("stage_type", "identity"),
+            desc="get_completed_stages",
+        ):
+            results[row["stage_type"]] = json.loads(row["identity"])
+
+        return results
+
+    async def set_session_data(self, session_id: str, key: str, value: Any):
```

Comment:
This is somewhat overly generalized, we only ever use this to set a single key. Do we really need this flexibility?

I also debated if it would be cleaner to store this data in a separate table, we would avoid having to update a JSON dict that way.

Reply:
yeah it might be over-general. I'm not sure I fancy ripping it out though. And in general json dicts work ok for storing extensible data in the database in a way that means we don't have to do migrations in future. I think it's ok as-is.

Reply:
Fair enough, just figured it was worth pointing out. I agree it is fine to leave it as is.